Google oauth login feature

tspenov opened the pull request:
**Done:**
- Implemented sign in with google for existing users, logging the last user in the db with equivalent email
- Used devise and omniauth-google-oauth2
- The google key and secret are in ENV var (_initializers/devise.rb_)

**Questions:**
- Should I register nonexistent users when signing in with google for the first time ?
- if yes: how should I fill city and login (_first_name.latinize.downcase ?_)
- should I write some tests ?

tspenov committed:
Added oauth google sign in

tspenov committed:
Add newer google-sign-in button
Remove signup from oauth
fix failure error messages

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,57 @@
+  	flash[:alert] = I18n.t "devise.omniauth_callbacks.failure", 
+  		:kind => "Google", :reason => reason
+  	redirect_to new_session_path
+  end
```
Please use the conventions in the project:
1. Two spaces for indentation
2. JSON-style hashes (`User.where(email: ...)` instead of `User.where(:email => ...)`) This is not is still not cleaned up everywhere, but all the new code is using it

tspenov replied:
It was properly indented in sublime but I'll check why it happens and fix it.

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,57 @@
+        unless user
+
+        # add provider and uid to the last user with this email
+        else
```
Please don't leave empty blocks like this. Just use `if user` and skip the `else` block

tspenov replied:
I'll fix this

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,57 @@
+      	#FIXME should I prepopulate email, name, login=given_name.downcase ?
+      	# delete the extra part of the hash to avoid cookie overflow
+      	request.env["omniauth.auth"].delete_if { |k,v| k =~ /\Aextra/ }
+        session["devise.google_data"] = request.env["omniauth.auth"]
```
I think this `session["devise.google_data"]` is not used anywhere. If that's the case, I think it should be removed

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,57 @@
+
+  def failure
+  	default_reason = "Комуникационна грешка"
+  	reason = default_reason
```
I think there is no real need for the `default_reason` variable here...

valo commented on app/views/layouts/main.html.erb:
```diff
@@ -24,8 +24,8 @@
-  <%- if flash[:error] -%>
-    <p class="error"><%= flash[:error] %></p>
+  <%- if flash[:error] or flash[:alert]-%>
+    <p class="error"><%= flash[:error] %><%= flash[:alert] %></p>
```
Is it possible to use the `flash[:error]` instead of introducing a new type of flash?

tspenov replied:
when I issued devise:install to initialize it said that it needed to handle notice and alert flashes from application.html.erb. But maybe in this usage of devise it is not used. I'll check this and remove it.

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,57 @@
+          user.save!
+        end
+
+      end
```
It seems to me that with the current logic an existing user will be able to login via google, but a non-existing user won't be able to register and login. Will it be possible to also register a new user if one is missing?

tspenov replied:
Yes I first made it to register but then removed it. How should I populate the city and the encrypted password when creating a user from google sign in ?

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,57 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
```
This class seems to lack specs. Will it be possible to have specs for it, please?

tspenov replied:
I'll add some specs

tspenov committed:
Register user if nonexistent
Fix indentation and comply to current conventions for hashes in the
project
Remove empty blocks
Remove devise.google_data from session because it is not used
Remove default_reason and use more common message catching all possible
failures
Use only flash[:error], remove :alert

tspenov committed:
Add google sign in feature test
Add omniauth callback controller test

tspenov committed:
Add more validation to omniauth_callbacks controller
Test the controller with valid and invalid data

tspenov commented:
Fixed the things in the comments and added controller and integration tests.

tspenov committed:
fix before repetition

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,73 @@
+      !auth.has_key?(:uid) or !auth.has_key?(:info) or
+      !auth[:info].has_key?(:email) or !auth[:info].has_key?(:name) or 
+      auth.uid.length == 0 or auth.provider.length == 0 or
+      auth.info.email.length == 0 or auth.info.email.length == 0
```
This condition looks really complex. I think it will be better if it is extracted into a separate private function

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,73 @@
+        user.role = User::CONTESTER
+        user.save! if user.valid?
+        register_user = true
+      end
```
This function is getting quite large. I think it can be improved if parts of it are extracted as separate private methods.

valo commented on config/routes.rb:
```diff
@@ -1,5 +1,6 @@
 MaycampArena::Application.routes.draw do
 
+  devise_for :users, :controllers => { :omniauth_callbacks => "users/omniauth_callbacks" }
```
Please use JSON style hashes (`omniauth_callbacks: ...`). This is the convention we try to keep around the project (when writing new code and refactoring old one).

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+    OmniAuth.config.add_mock(:google_oauth2, mock_auth_hash())
+    request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:google_oauth2]
+  
+    @user_with_uid_provider = {
```
Using `let` to define such variables is much better as they will be evaluated lazily. Like this:

``` ruby
let(:user_with_uid_provider) do
 {
  ...
 }
end
```

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+
+  before(:each) do
+    
+    request.env["devise.mapping"] = Devise.mappings[:user]
```
Why is this needed here? I don't see it used anywhere

tspenov replied:
we need to explicitly tell Devise which mapping to use because in testing
we are bypassing the router

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+    
+    request.env["devise.mapping"] = Devise.mappings[:user]
+    OmniAuth.config.add_mock(:google_oauth2, mock_auth_hash())
+    request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:google_oauth2]
```
These here can probably be extracted as a function similarly to `create_invalid_mock_hash`. Then there could be 2 context blocks for the valid and invalid authentication blocks. Like this:

``` ruby
context 'with valid oauth authentication' do
  before do
    create_valid_mock_hash
  end
  ...
end

context 'with invalid oauth authentication' do
  before do
    create_invalid_mock_hash
  end
  ...
end

```

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+          post :google_oauth2
+        }.to_not change{ User.count }
+
+        
```
The whitespace is a kind of inconsistent around this file. In some places there are 2 empty lines. In other places there is 1 empty line. Can the white space be unified? Generally I believe the `it` and `context` blocks should be separated with one empty line. You can have a look at the convention followed in the other specs in the project.

valo commented on spec/features/google_sign_in_spec.rb:
```diff
@@ -0,0 +1,113 @@
+    expect(page).to have_text("Не успяхме да ви оторизираме чрез Google")
+  end
+
+end
```
Generally the project uses `cucumber` for integration tests. You can have a look at https://github.com/tspenov/maycamp_arena/tree/master/features. If you feel like it you can try to use that for integration testing. Also please make sure the cucumber features are passing

valo commented on spec/rails_helper.rb:
```diff
@@ -50,4 +52,6 @@
   config.include AutenticationHelpers, type: :controller
   config.include ActionDispatch::TestProcess
+
+  OmniAuth.config.test_mode = true
```
I think this config should be added in the https://github.com/tspenov/maycamp_arena/tree/master/spec/support folder with a proper name (`omniauth.rb`?)

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+        create_invalid_mock_hash(opts)
+        post :google_oauth2
+        expect(flash[:error]).to be_present
+        expect(flash[:error]).to eq "Не успяхме да ви оторизираме чрез Google поради невалидни данни"
```
These 2 lines are redundant. One of them will be enough.

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+        User.create(@user_with_email)
+        post :google_oauth2
+        expect(flash[:notice]).to be_present
+        expect(flash[:notice]).to eq "Успешно оторизиран чрез Google профил."
```
These 2 lines are redundant

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+      role: "contester",
+      name: "Test user",
+      city: "Айтос"
+    }
```
Isn't it possible to use `attributes_for(:user)` here?

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+      city: "Айтос"
+    }
+
+    User.delete_all
```
Why is that needed? The DB should be reset after each spec.

valo commented on spec/controllers/users/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,170 @@
+          post :google_oauth2
+        }.to change{ User.count }.by(1)
+
+        user = User.last
```
A more reliable way to handle this might be `User.find_by(email: mock[:info][:email])`

tspenov committed:
Modify controller and isolate some of the logic in private functions
Fix non json style hash in routes.rb
add common omniauth test config in separate support file
Fix inconsistency in newlines
Use let where possible
Use attribute_for(:user) where posssible
Remove redundant flash checks
Fix context messages using with,when
Remove User.delete_all because not needed
Substitute User.last with User.find_by...

tspenov committed:
Remove old integration test
Add cucumber integration test

tspenov committed:
Add Background for repeating information
Add And for readability

tspenov committed:
Fix table indentation

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,85 @@
+      user.password = Devise.friendly_token[0,20] #??? should I fill password field or skip it
+      user.name = auth.info.name
+      user.login = auth.uid   #??? should I generate some other uniq random string for login
+      user.role = User::CONTESTER
```
There is an easier way to make such big assignments:

``` ruby
user.update(
  provider: ...,
  uid: ...,
  email: ...,
)
```

valo commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,85 @@
+      !auth.has_key?(:uid) or !auth.has_key?(:info) or
+      !auth[:info].has_key?(:email) or !auth[:info].has_key?(:name) or 
+      auth.uid.length == 0 or auth.provider.length == 0 or
+      auth.info.email.length == 0 or auth.info.email.length == 0
```
There is a shorter way to express this:

``` ruby
!auth || auth.provider.blank? || auth.uid.blank? || auth.info.blank? || auth.info.name.blank? || auth.info.email.blank?
```

valo commented:
Apart from the small cosmetic changes I think we are good to go :) :+1:

tspenov committed:
Simplify complex conditions and assignments
